### DISCUSSION TITLE
how to enable two SPA to share the same origin in the local development environment using the webpack dev server

### DISCUSSION THREAD

**Discussion by joaomelo** [webpack/webpack#17827] [Q&A]
I have two web single-page applications that are served independently. They share the same origin and are routed by path. The web-auth app is served at http://www.some-app.com/web-auth, and the web-intra is served at http://www.some-app.com/web-intra.

In our production environment, all is done by the enterprise gateway software and is configured in an interface provided to devs. I want to emulate the same behavior in the local environment. 

The web-auth app is served with webpack dev server. Here is the app code:
```js
//index.js
const container = document.createElement("div");

const text = document.createElement("p");
text.innerHTML = "this is the auth page";
container.appendChild(text);

const input = document.createElement("input");
container.appendChild(input);

const button = document.createElement("button");
button.innerHTML = "go";
button.onclick = () => {
  const userData = { email: "email@email.com", token: input.value };
  localStorage.setItem("user", JSON.stringify(userData));
};
container.appendChild(button);
document.body.appendChild(container);
```

Here is the webpack configuration for web-auth:

```js
const path = require("path");
const HtmlWebpackPlugin = require("html-webpack-plugin");

module.exports = {
  mode: "development",
  entry: "./src/index.js",
  devtool: "inline-source-map",
  devServer: {
    historyApiFallback: true,
    devMiddleware: { index: false },
    static: "./dist",
    port: 9001,
  },
  plugins: [new HtmlWebpackPlugin({ title: "auth" })],
  output: {
    filename: "[name].bundle.js",
    path: path.resolve(__dirname, "dist"),
    clean: true,
  },
  optimization: {
    runtimeChunk: "single",
  },
};
```

The web-intra app is almost identical but is served on port 9002.

Both web-auth and web-intra are confirmed to run in the local environment if accessed at http://localhost:9001 and http://localhost:9002.

To create the same origin effect, I raised a third web-proxy app with a dummy entry point:

```js
//index.js
const container = document.createElement("div");

const text = document.createElement("p");
text.innerHTML = "this is the proxy page";
container.appendChild(text);

document.body.appendChild(container);
```

The web-proxy webpack configuration attempts to do the magic with:

```js
const path = require("path");
const HtmlWebpackPlugin = require("html-webpack-plugin");

module.exports = {
  mode: "development",
  entry: "./src/index.js",
  devtool: "inline-source-map",
  devServer: {
    static: "./dist",
    host: "localhost",
    port: 9000,
    open: false,
    devMiddleware: { index: false },
    proxy: {
      "/web-auth": {
        target: "http://localhost:9001",
        // target: "https://www.google.com",
        changeOrigin: true,
        pathRewrite: () => "",
      },
      "/web-intra": {
        target: "http://localhost:9002",
        // target: "https://www.google.com",
        changeOrigin: true,
        pathRewrite: () => "",
      },
    },
  },
  plugins: [new HtmlWebpackPlugin({ title: "proxy" })],
  output: {
    filename: "[name].bundle.js",
    path: path.resolve(__dirname, "dist"),
    clean: true,
  },
  optimization: { runtimeChunk: "single" },
};
```

But this doesn't work. I've read the webpack dev server docs and various articles on the web. Most of the content is focused on tunneling requests to some backend api which is not the problem here. 

[This article](https://dev.to/0916dhkim/about-reverse-proxy-312l) attempts something similar to my environment, but the configurations I ported to my solution weren't sufficient to solve the problem.

If I change the target option to an external site, it works, and the proxy app does its proxying to the external site without changing the browser URL, but when the target is set to localhost, the proxying does not work.

Here is the failure:

![image](https://github.com/webpack/webpack/assets/5727050/76d48e38-d448-44ca-9b0e-319432b66d51)

I appreciate the help.

**Comment by alexander-akait** [comment]
@joaomelo I think something wrong because you use localhost (we need to look at proxy middleware code), did you try to use `/etc/hosts` and use any custom host to check will it works? Sorry for delay

**Reply by joaomelo** [reply]
Hi @alexander-akait. How are you?

I didn't try to do any customization on the host machine. This solution was intended to be used by multiple devs using different OS. I attempted to deliver a dev environment where they could hit an npm script and start developing right away.

Thank you very much for reaching out.
